src/base/encoding.rs

```diff
@@ -1,285 +1,280 @@
-    &encoding_rs::ISO_2022_JP_INIT,
-    &encoding_rs::REPLACEMENT_INIT,
-    &encoding_rs::REPLACEMENT_INIT,
-    &encoding_rs::REPLACEMENT_INIT,
-    &encoding_rs::WINDOWS_1250_INIT,
-    &encoding_rs::WINDOWS_1251_INIT,
-    &encoding_rs::WINDOWS_1252_INIT,
-    &encoding_rs::WINDOWS_1253_INIT,
-    &encoding_rs::WINDOWS_1254_INIT,
-    &encoding_rs::WINDOWS_1255_INIT,
-    &encoding_rs::WINDOWS_1256_INIT,
-    &encoding_rs::WINDOWS_1257_INIT,
-    &encoding_rs::WINDOWS_1258_INIT,
-    &encoding_rs::ISO_8859_6_INIT,
-    &encoding_rs::ISO_8859_8_INIT,
-    &encoding_rs::ISO_8859_6_INIT,
-    &encoding_rs::ISO_8859_8_I_INIT,
-    &encoding_rs::ISO_8859_7_INIT,
-    &encoding_rs::EUC_KR_INIT,
-    &encoding_rs::UTF_8_INIT,
-    &encoding_rs::UTF_8_INIT,
-    &encoding_rs::EUC_KR_INIT,
-    &encoding_rs::WINDOWS_1252_INIT,
-    &encoding_rs::EUC_KR_INIT,
-    &encoding_rs::X_MAC_CYRILLIC_INIT,
-    &encoding_rs::X_USER_DEFINED_INIT,
-    &encoding_rs::GBK_INIT,
-    &encoding_rs::UTF_16LE_INIT,
-    &encoding_rs::WINDOWS_1252_INIT,
-    &encoding_rs::ISO_8859_2_INIT,
-    &encoding_rs::ISO_8859_6_INIT,
-    &encoding_rs::ISO_8859_7_INIT,
-    &encoding_rs::ISO_8859_3_INIT,
-    &encoding_rs::ISO_8859_4_INIT,
-    &encoding_rs::ISO_8859_5_INIT,
-    &encoding_rs::ISO_8859_8_INIT,
-    &encoding_rs::UTF_8_INIT,
-    &encoding_rs::WINDOWS_1254_INIT,
-    &encoding_rs::ISO_8859_7_INIT,
-    &encoding_rs::X_MAC_CYRILLIC_INIT,
-    &encoding_rs::REPLACEMENT_INIT,
-    &encoding_rs::ISO_8859_6_INIT,
-    &encoding_rs::ISO_8859_8_INIT,
-    &encoding_rs::UTF_8_INIT,
-    &encoding_rs::ISO_8859_5_INIT,
-    &encoding_rs::EUC_JP_INIT,
-];
-
-fn encoding_to_index(encoding: AsciiCompatibleEncoding) -> usize {
-    let encoding: &'static Encoding = encoding.into();
-
-    ALL_ENCODINGS
-        .iter()
-        .position(|&e| e == encoding)
-        .expect("the ALL_ENCODINGS is not complete and needs to be updated")
-}
-
 /// A charset encoding that can be shared and modified.
 ///
 /// This is, for instance, used to adapt the charset dynamically in a [`crate::HtmlRewriter`] if it
 /// encounters a `meta` tag that specifies the charset (that behavior is dependent on
 /// [`crate::Settings::adjust_charset_on_meta_tag`]).
 #[derive(Clone)]
 pub struct SharedEncoding {
-    encoding: Arc<AtomicUsize>,
+    encoding: Arc<AtomicPtr<Encoding>>,
 }
 
 impl SharedEncoding {
     #[must_use]
     pub fn new(encoding: AsciiCompatibleEncoding) -> Self {
+        let encoding: &'static Encoding = encoding.into();
         Self {
-            encoding: Arc::new(AtomicUsize::new(encoding_to_index(encoding))),
+            // *mut T doesn't have aliasing requirements
+            encoding: Arc::new(AtomicPtr::new(ptr::from_ref(encoding).cast_mut())),
         }
     }
 
     #[must_use]
     pub fn get(&self) -> &'static Encoding {
         let encoding = self.encoding.load(Ordering::Relaxed);
-        ALL_ENCODINGS[encoding]
+        unsafe { &*encoding }
```

Comment:
Can we not have `unsafe` blocks for that? I'm fine with a lookup table as a price for that

Reply:
What's wrong with that?

The pointers have a static lifetime, so it's not a risky code.